[16.0][ADD] repair_preparation

sbejaoui opened the pull request:
The repair module consumes spare parts at the end of the repair without considering the qunatity reservation. the more natural flow is to have a **staging step** to bring and reserve parts **before** the repair begins, improving planning and reducing delays at the workbench.

This addon introduces a **Preparation** flow:

-  On **Confirm** (validate) of the repair, a procurement is run procurement for all eligible **Add** lines
-  When a repair line is **created/edited** while the repair is *Under Repair*:

   -  If any linked move is **Done** → editing raises a validation error
   -  Otherwise, linked moves are **canceled** and procurement is **re-run** for the updated lines

-  **Finishing** a repair is blocked if:

   -  There are consumed lines but **no preparation pickings**, or
   -  Preparation pickings exist but are **not done**

cc/ @lmignon , @rousseldenis

sbejaoui committed:
[16.0][ADD] repair_preparation

rousseldenis commented on repair_preparation/readme/DESCRIPTION.md:
```diff
@@ -0,0 +1,12 @@
+The repair module consumes spare parts at the end of the repair without considering the qunatity reservation.
```
```suggestion
The repair module consumes spare parts at the end of the repair without considering the quantity reservation.
```

rousseldenis commented on repair_preparation/models/res_company.py:
```diff
@@ -0,0 +1,23 @@
+        "repairs can be finished without preparation checks.",
+    )
+
+    repair_preparation_picking_type_id = fields.Many2one(
```
Couldn't have this be replaced by a configuration on warehouse level to enable repair preparations (like in core)?

sbejaoui replied:
yes good point, it should be at warehouse level, not the company

rousseldenis reviewed (commented):
LGTM. Some comments

sbejaoui committed:
[IMP] repair_preparation: update repair line after preparation move done

update repair line location and lot after preparation is done

sbejaoui committed:
[FIX] repair_preparation: preparation_picking_type_id is required if repair_preparation enabled

rousseldenis commented on repair_preparation/models/repair_line.py:
```diff
@@ -0,0 +1,66 @@
+                        "because some linked moves are already done.\n"
+                        "Repair: %(repair)s\nLines: %(lines)s"
+                    )
+                    % {
```
Please use translation function arguments instead.

rousseldenis commented on repair_preparation/models/repair_order.py:
```diff
@@ -0,0 +1,158 @@
+        if not self.preparation_group_id:
+            self.preparation_group_id = self.env["procurement.group"].create(
+                {
+                    "name": _("Preparation for") + self.name,
```
```suggestion
                    "name": _("Preparation for ") + self.name,
```

rousseldenis commented on repair_preparation/models/repair_order.py:
```diff
@@ -0,0 +1,158 @@
+        return operations.filtered(
+            lambda line: line.type == "add"
+            and line.product_id
+            and line.product_uom_qty > 0
```
No need of `float_compare()` ?

sbejaoui committed:
[IMP] repair_preparation: use repair_warehouse to define warehouse link

sbejaoui committed:
[DON'T MERGE] add unmerged requirements

sbejaoui committed:
[IMP] repair_preparation: skip preparation flow for customer location repairs

nicolas-delbovier-acsone committed:
[FIX] repair_preparation: cancel preparation picking if repair order when cancelled

nicolas-delbovier-acsone committed:
[IMP] repair_preparation: add count and invisibility on smart buttons

nicolas-delbovier-acsone committed:
[IMP] repair_preparation: general code improvements

- Do not call action on empty recordset
- Remove depends in manifest
- Use trnaslate function args

nicolas-delbovier-acsone committed:
[FIX] repair_preparation: prevent installation with repair_stock_move

Problem:
The 'repair_stock_move' module overrides standard behavior in a way that
causes duplicate stock moves when used alongside this module. Because
this behavior is hard-coded and not toggleable, there is no way to
prevent the conflict via configuration.

nicolas-delbovier-acsone commented:
I think there is an incompatibility between `repair_refurbish` and ` repair_type` because of this code in `repair_type`:
```python
@api.onchange("type")
def onchange_operation_type(self):
    # this onchange was overriding the changes from the compute
    # method `_compute_location_id`, we ensure that the locations
    # in the types have more priority by explicit calling the compute.
    res = super().onchange_operation_type()
    self._compute_location_id()   # <- THIS LINE IS THE CULPRIT
    return res
```

The logic for the computation of the locations are not compatible.

**NB:** I tested on a database with only  `repair_refurbish` and ` repair_type` modules installed and the unit tests also fail

nicolas-delbovier-acsone committed:
[FIX] repair_preparation: use `float_compare` for float comparison

nicolas-delbovier-acsone committed:
[IMP] repair_preparation: use repair location for preparation pickings

Before this commit, the stock picking generated for preparing spare parts
used the default destination location defined on the picking type.

This resulted in poor UX, as the operator was forced to manually edit the
destination location on the generated picking to match the specific
location of the repair order before validating the transfer. This manual
step was confusing, added unnecessary clicks, and was prone to errors.

Now, the procurement uses the `location_id` of the Repair Order as its
destination. The generated picking is automatically routed to the exact
location where the repair takes place, eliminating the need for manual
intervention and streamlining the operator's workflow.

nicolas-delbovier-acsone committed:
[IMP] repair_preparation: use repair location for preparation pickings

Before this commit, the stock picking generated for preparing spare parts
used the default destination location defined on the picking type.

This resulted in poor UX, as the operator was forced to manually edit the
destination location on the generated picking to match the specific
location of the repair order before validating the transfer. This manual
step was confusing, added unnecessary clicks, and was prone to errors.

Now, the procurement uses the `location_id` of the Repair Order as its
destination. The generated picking is automatically routed to the exact
location where the repair takes place, eliminating the need for manual
intervention and streamlining the operator's workflow.

nicolas-delbovier-acsone committed:
[FIX] repair_preparation: allow ending repair with cancelled preparation pickings

nicolas-delbovier-acsone committed:
[IMP] repair_preparation: lock operation lines when confirmed for sync integrity

When the repair preparation flow is enabled, procurement moves are generated
upon order confirmation or when "Under Repair". Modifying quantities or
products while the order is in the 'Confirmed' state (but before the
repair has started) can lead to stock moves getting out of sync with the
repair lines.

rousseldenis commented:
@sbejaoui @nicolas-delbovier-acsone Tests seem red

nicolas-delbovier-acsone commented:
> @sbejaoui @nicolas-delbovier-acsone Tests seem red

Yes I know, as I said above, installing **ONLY** ` repair_refurbish` and `repair_type` on a test database on this branch makes tests fail for some reason... I do not know if this is an issue with the current module or if it is rooted somewhere else.

Maybe rebasing the branch will fix it? I'll test that when I have more time

nicolas-delbovier-acsone committed:
[REF] repair_preparation: add args name in named tuple for clarity

nicolas-delbovier-acsone committed:
[IMP] repair_preparation: improve preparation picking view action

Previously, the module used a standard stock action, which led to several
usability issues:
- Breadcrumb name "Transfers" was inconsistent with the "Preparation" context.
- The "New" button was visible despite creation being invalid in this flow.
- Users were forced through a list view even when only one picking existed.

This change refactors the action to be generated dynamically in Python. It:
- Sets the breadcrumb title to "Preparation Pickings".
- Disables creation via the action context.
- Automatically opens the form view if only one record is found.
